torax/sources/electron_cyclotron_source.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import annotations
+
+import dataclasses
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from torax import array_typing
+from torax import geometry
+from torax import interpolated_param
+from torax import state
+from torax.config import runtime_params_slice
+from torax.constants import CONSTANTS
+from torax.sources import runtime_params as runtime_params_lib
+from torax.sources import source
+
+
+@dataclasses.dataclass(kw_only=True)
+class RuntimeParams(runtime_params_lib.RuntimeParams):
+    """Runtime parameters for the electron-cyclotron source."""
+
+    # Global dimensionless current drive efficiency
+    global_efficiency: (
+        runtime_params_lib.interpolated_param.InterpolatedVarSingleAxisInput
+    )
+
+    # EC power density profile on the rho grid; units [W/m^2]
+    ec_power_density: runtime_params_lib.interpolated_param.InterpolatedVarTimeRhoInput
```

Comment:
May be nice to have options on how to set the power density. This one is good as a prescribed option. Can maybe change the name to: "prescribed_ec_power_density" ?

Could think of having Gaussian parameters (width, center, totalpower) which are TimeInterpolated runtime_params. Then maybe if "prescribed_ec_power_density" is None then these Gaussian ones are used, and if it isn't none then the prescribed_ec_power_density profile is used?

Reply:
Sorry I missed earlier the comment about the Gaussian option.

I think it's fine to evaluate the Gaussian at each evaluation. We already do this for the generic_ion_el_heat_source and nbi_particle_source (soon to be renamed generic_particle_source) and pellet_source. It's jitted and the Gaussian evaluation time is negligible.

A pattern which you can perhaps reuse is that of the pellet source. See https://github.com/google-deepmind/torax/blob/b3b3692b8491c9fa86d6229975afb943b641bc3c/torax/sources/electron_density_sources.py#L269 . Note that it uses `formulas.gaussian_profile` which is a reusable function. Note that this is a FORMULA_BASED source, but you can ignore that. We're going to drop FORMULA_BASED and just make it a special case of MODEL_BASED. You can use the `formula.gaussian_profile` function as part of the model_func too of course.

By a sum-of-Gaussians, do you mean that you'd like a model where different antennas will set different Gaussian depositions at different radii, and each Gaussian will need its own set of time-dependent parameters? If so, then if it's not too many Gaussians to be summed then I guess you can give each one it's own labelling in the runtime_params, and call `formulas.gaussian_profile` for each one and sum them. Another thing we could consider, which would require a bit of feature extension, is for the TimeInterpolatedInput to allow a tuple of parameters at each timeslice, and get_value time interpolates each tuple element (and this could be vectorized), and formulas.gaussian_profile is also extended for tuple input, returning the sum of Gaussians. @Nush395 what do you think of feasibility of this last point?

Reply:
Yes, that's absolutely right - N antenna whose deposition can each be modelled as a Gaussian at a specific radius + width/height and the total EC heating profile as a sum of the Gaussian profiles.
Different machines might have different numbers of EC antennae (e.g. STEP will have something like 5-20), so I was thinking that if a Gaussian-based profile was required then it would be best to support an arbitrary number of antennae, rather than 1 or a predefined number.

Reply:
Happy to contribute to the feature, but I wonder whether it would be better suited as a separate PR. Could do the single Gaussian profile in this PR, if that would be preferable, but if it's going to be superseded by the sum-of-Gaussians change I'm not sure how worthwhile it is.

Reply:
I think a single Gaussian profile is fine in this PR, using formulas.gaussian_profile. In a subsequent one we can extend TimeInterpolatedInput and formulas.gaussian_profile to then enable a sum of Gaussians. It's an extension rather than a superseding, so worthwhile to do it staged.